fix: Create only one instance of `StatusChatInput`

igor-sirotin opened the pull request:
Fixes https://github.com/status-im/status-desktop/issues/10436

## Changes description

* Remove #10379 hotfix
* Moved `StatusChatInput` to `ChatColumnView`, so that it's only created once
* Added storage to `chat_content/input_area` to preserve these properties for each chat:
    * text
    * reply message id (new feature 🤘 )
    * attached images (new feature 🤘)
* Create popups, dialogs and menus dynamically

## Performance improvement

Test description:
1. Launch the app
2. Login with `status-shared-account`
3. Land on status community (as selected on previous launch)

| Description | Commit | RAM usage, Gb |
|-|-|-|
| Before changes, with #10379 hotfix | ee9db6a3024780c8fcfa4f7687af18b800f4ebd1 | 0.83 Gb |
| Before changes, without hotfix | 5ca6e7de93b18ef79a1f6f974bbc983e256da999 | 1.52 Gb |
| After changes | e45734320221161cf3700ea9a4307c1af41046ab | 0.80 Gb |

## Affected areas

chat

## Screenshot

Input preserved state demo:

https://github.com/status-im/status-desktop/assets/25482501/31ce52fa-df5f-4333-8d0d-6a613ac3f812

igor-sirotin commented on src/app/modules/main/chat_section/chat_content/input_area/model.nim:
This was simply not used

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | e4573432 | [#1](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-imports/job/PR-10928/1/) | 2023-06-05 14:08:38 | ~5 min | `tests/imports` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-imports/job/PR-10928/1//consoleText) |

caybro commented on ui/app/AppLayouts/Chat/views/ChatView.qml:
```diff
@@ -89,9 +89,6 @@ StatusSectionLayout {
             Global.openPopup(statusStickerPackClickPopup, {packId: stickerPackId, store: root.stickersPopup.store} )
         }
-        onOpenAppSearch: {
-            root.openAppSearch();
```
Not needed any longer?

igor-sirotin replied:
Well it just wasn't ever called from `ChatContenView` 🙂

jrainville commented:
> @igor-sirotin potentially fixing this too? #10885

Yeah I saw the code change and immediately thought that it would probably fix it

jrainville reviewed (commented):
Great job. 

Some small comments only.

igor-sirotin commented on src/app/modules/main/chat_section/chat_content/input_area/model.nim:
🤷

igor-sirotin commented on ui/app/AppLayouts/Chat/views/ChatColumnView.qml:
```diff
@@ -107,46 +201,160 @@ Item {
+                id: chatRepeater
+                model: parentModule && parentModule.model
+
+                ChatContentView {
```
There's a `Loader` inside the `ChatContentView`, which is controlled with `chatMessagesLoader.active: model.loaderActive`. The one that makes caching 5 chats.

igor-sirotin commented on ui/app/AppLayouts/Chat/views/ChatColumnView.qml:
```diff
@@ -107,46 +201,160 @@ Item {
+                d.activeChatContentModule.inputAreaModule.preservedProperties.fileUrlsAndSourcesJson = JSON.stringify(chatInput.fileUrlsAndSources)
+            }
+
+            // TODO: A lot of chatContentModule.getMyChatId is used here. Isn't it just `root.activeChatId`?
```
I will remove it, as it won't give any benefits. Only cleanup and potential bugs :D

igor-sirotin commented:
> @igor-sirotin potentially fixing this too? #10885

Tested. Unfortunately, it doesn't fix the issue 🙁

osmaczko reviewed (approved):
Good job overall. While it's not my preference to store QML component states in the backend, I understand the constraints that led to this approach given the complexity of the `StatusChatInput` component. I'm okay with proceeding as is.

osmaczko commented:
If these changes were a milk carton, they'd have 'Missing: QA' on the side :laughing:

caybro commented:
@igor-sirotin the `e2e` test failure looks legit and will need fixing

igor-sirotin commented:
> While it's not my preference to store QML component states in the backend

@osmaczko I agree and tried to implement this in QML at first. But the logic was quite confusing, I would have to create a `Map` holding info for each chat. And having this in backend, we have an ability later to keep these message drafts in database, or even sync between devices (ofc I'm not sure if it's ever needed 😄).

igor-sirotin commented on ui/app/AppLayouts/Chat/views/ChatColumnView.qml:
```diff
@@ -98,6 +99,99 @@ Item {
+
+    Component.onCompleted: {
+        d.fixUsersStore()
+        d.fixMessageStore()
```
dropped with usage of `Qt.callLater`

igor-sirotin commented:
@anastasiyaig indeed, this would probably be good to test.
Can you please pick it up when there's time? 🙂 

#### What to test:
1. Chat input area basic functionality:
    - sending messages
    - opening popups, their position (gif, emoji, stickers, text styling)
    - attaching images, removing attached images
2. Input area preserves state for each chat when switching between chats:
     - text
     - images
     - reply

anastasiyaig commented:
@igor-sirotin sure, it will be tested

glitchminer commented:
Hi @igor-sirotin , QA results below. One query came up - is there still a limit on how many channels messages will persist in?

---
1. **OPEN** When gif selector is open it does not toggle closed when button is clicked again

https://github.com/status-im/status-desktop/assets/50769329/013c6747-2b4c-4b82-ad5b-102e627d382d

---
2. **OPEN** Alignment of added sticker packs is different to 0.12.0
![image](https://github.com/status-im/status-desktop/assets/50769329/8050a943-1329-4be7-9513-ef44e47de022)

---
Summary:
3 users + community with 10 channels

- Last open channel returned to on restart
- Send message, image, reply
- Search for emoji
- Enable Tenor, search for gif, send
- Add stickers and send
- Text formatting popup opens when message is highlighted
- Images preserved when switching channels
- Reply preserved when switching channels
- Reply with images preserved when switching channels
- Adding 5 images, removed 2, added 1 - preserved and sent OK.

igor-sirotin commented:
@glitchminer thanks for testing! Fixed both points 👌 

As for a limit on channels persisted:
That's true, but only for chats themselves. So that last 5 chats would load fast when switching. But the input area preserves draft messages for all chats. Because that's not that much information comparing to chat components.

glitchminer commented:
Thanks @igor-sirotin , those 2 are resolved now. Also sorry, I did just find one more regression just now related to the gif and sticker popups.

- Open stickers
- Open gifs
- Close gifs
- Stickers is still open

https://github.com/status-im/status-desktop/assets/50769329/e0de1222-33ef-4afd-a3f1-28ed1a0cc2f4

igor-sirotin commented:
> * Open stickers
> * Open gifs
> * Close gifs
> * Stickers is still open

@glitchminer indeed, thanks for that!

I've fixed and simplified the logic for opening/closing emoji, gif and stickers popup. Should be fine by now.

Also, now clicking on anything inside the input area (text area, buttons, whatever) will close the popup as well. This changes the behaviour to discord-style. Previously it was possible to type the message in and still see the emoji/gif/stickers popup.

https://github.com/status-im/status-desktop/assets/25482501/b8a56d15-4ead-433c-b2b3-411832486dc4

glitchminer commented:
Great, thanks @igor-sirotin. That's done now.
